File: src/contests/cg/puzzle/swift/telephone-numbers.swift
// 🎮 CodinGame Puzzle - telephone-numbers
// https://www.codingame.com/training/medium/telephone-numbers

let n = Int(readLine()!)!
// Trie as a flat array of nodes, each a digit -> child-index map.
var nodes: [[Character: Int]] = [[:]]
var cable = 0
for _ in 0..<n {
    let number = Array(readLine()!)
    var cur = 0
    for c in number {
        if nodes[cur][c] == nil {
            let id = nodes.count
            nodes[cur][c] = id
            nodes.append([:])
            cable += 1
        }
        cur = nodes[cur][c]!
    }
}
print(cable)

File: src/contests/cg/puzzle/rust/telephone-numbers.rs
// 🎮 CodinGame Puzzle - telephone-numbers
// https://www.codingame.com/training/medium/telephone-numbers

use std::collections::HashMap;

fn rl() -> String {
    let mut s = String::new();
    std::io::stdin().read_line(&mut s).unwrap();
    s.trim().to_string()
}

fn main() {
    let n: usize = rl().parse().unwrap();
    // Trie as a flat vec of nodes, each a digit -> child-index map.
    let mut nodes: Vec<HashMap<char, usize>> = vec![HashMap::new()];
    let mut cable = 0;
    for _ in 0..n {
        let number = rl();
        let mut cur = 0;
        for c in number.chars() {
            if !nodes[cur].contains_key(&c) {
                let id = nodes.len();
                nodes[cur].insert(c, id);
                nodes.push(HashMap::new());
                cable += 1;
            }
            cur = nodes[cur][&c];
        }
    }
    println!("{}", cable);
}
